File: src/learning_tf/src/dynamic_tf_broadcaster.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/LinearMath/QuadWord.hpp"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "tf2_ros/transform_broadcaster.hpp"
#include <memory>
#include <chrono>

using namespace std::chrono_literals;

class DynamicTFBroadcasterNode: public rclcpp::Node
{
private:
    std::shared_ptr<tf2_ros::TransformBroadcaster> broadcaster_;
    rclcpp::TimerBase::SharedPtr timer_;
public:
    DynamicTFBroadcasterNode() : Node("dynamic_tf_broadcaster_cpp")
    {
        broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
        timer_ = this->create_wall_timer(100ms, std::bind(&DynamicTFBroadcasterNode::time_callback, this));
    }

    void time_callback()
    {
        geometry_msgs::msg::TransformStamped trans;
        trans.header.stamp = this->get_clock()->now();
        trans.header.frame_id = "map";
        trans.child_frame_id = "base_link";
        trans.transform.translation.x = 2.0;
        trans.transform.translation.y = 3.0;
        trans.transform.translation.z = 0.0;
        tf2::Quaternion q;
        q.setRPY(0.0, 0.0, 30 * M_PI / 180.0);
        trans.transform.rotation = tf2::toMsg(q);
        broadcaster_->sendTransform(trans);
    }
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DynamicTFBroadcasterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();

    return 0;
}

File: src/learning_tf/scripts/dynamic_tf_broadcaster.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from tf2_ros import StaticTransformBroadcaster # 静态坐标发布器
from geometry_msgs.msg import TransformStamped # 消息接口
from tf_transformations import quaternion_from_euler # 欧拉角转四元数
import math

class DynamicTFBroadcaster(Node):
    def __init__(self):
        super().__init__('dynamic_tf_broadcaster')
        self.static_broadcaster_ = StaticTransformBroadcaster(self)
        self.timer_ = self.create_timer(1.0, self.publish_dynamic_tf)
        
    def publish_dynamic_tf(self):
        """
        发布TF camera_link 到 bottle_link的关系
        """
        transform = TransformStamped()
        transform.header.frame_id = 'camera_link'
        transform.child_frame_id = 'bottle_link'
        transform.header.stamp = self.get_clock().now().to_msg()
        
        transform.transform.translation.x = 0.2
        transform.transform.translation.y = 0.3
        transform.transform.translation.z = 0.5

        # 欧拉角转四元数 q = x, y, z, w
        q = quaternion_from_euler(0, 0, 0)
        transform.transform.rotation.x = q[0] 
        transform.transform.rotation.y = q[1] 
        transform.transform.rotation.z = q[2] 
        transform.transform.rotation.w = q[3] 
        
        self.static_broadcaster_.sendTransform(transform)
        self.get_logger().info(f'发布TF:{transform}')

def main():     
    rclpy.init()
    node = DynamicTFBroadcaster()
    rclpy.spin(node)
    rclpy.shutdown()
    
if __name__ == '__main__':
    main()
